src/main/java/com/palantir/refreshable/DefaultRefreshable.java

```diff
@@ -159,14 +148,57 @@ private void preSubscribeLogging() {
         }
     }
 
+    @Override
+    public <R> Refreshable<R> map(Function<? super T, R> function) {
+        readLock.lock();
+        try {
+            R initialChildValue = function.apply(current);
+            SettableRefreshable<R> child = createSettableChild(initialChildValue);
+
+            // Contract for Cleaner#register states that the Runnable should never reference the object being registered
+            Disposable disposable = registerMapSubscriber(function, new WeakReference<>(child));
+
+            REFRESHABLE_CLEANER.register(child, disposable::dispose);
+
+            return child;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    private <R> SettableRefreshable<R> createSettableChild(R initialChildValue) {
+        Optional<?> parentReference = Optional.of(this);
+        return new DefaultRefreshable<>(initialChildValue, parentReference, rootSubscriberTracker);
+    }
+
+    private <R> Disposable registerMapSubscriber(
+            Function<? super T, R> function, WeakReference<SettableRefreshable<R>> childRef) {
+        MappingSubscriber<? super T, R> mapSubscriber = new MappingSubscriber<>(function, childRef);
+        return subscribeToSelf(mapSubscriber, _ignored -> {});
+    }
+
     @GuardedBy("readLock")
-    private Disposable subscribeToSelf(Consumer<? super T> subscriber) {
+    private <C extends NoThrowSubscriber<? super T>> Disposable subscribeToSelf(
+            C subscriber, DisposeListener<C> disposeListener) {
         preSubscribeLogging();
+        // The subscriber promises not to throw, so there isn't a way we could have a hanging
         orderedSubscribers.add(subscriber);
         subscriber.accept(current);
-        return new DefaultDisposable(orderedSubscribers, subscriber);
+        return new NotifyingDisposable<>(
+                subscriberToRemove -> {
+                    orderedSubscribers.remove(subscriberToRemove);
+                    disposeListener.disposed(subscriberToRemove);
+                },
+                subscriber);
     }
 
+    interface DisposeListener<C extends Consumer<?>> {
+        void disposed(C consumer);
+    }
+
+    // Marker interface for documentation: if you implement this, you must promise not to throw!
+    interface NoThrowSubscriber<T> extends Consumer<T> {}
```

Comment:
Should we use a concrete delegating type which catches exceptions instead? That way we don't have to trust that implementations follow the contract -- they're forced to. It means we can't accidentally use method references to convert unsafe consumers.

Reply:
Yes, except that existing ones already catch. But that works for me even more.